addons/web/static/tests/views/fields/one2many_field_tests.js

```diff
@@ -14509,4 +14509,141 @@ QUnit.module("Fields", (hooks) => {
+            serverData.models.partner.onchanges = {
+                bar: (obj) => {
+                    obj.turtles = [[0, false, { display_name: "onchange name" }]];
+                },
+            };
+
+            await makeView({
+                type: "form",
+                resModel: "partner",
+                serverData,
+                arch: `
+                    <form>
+                        <field name="bar"/>
+                        <field name="turtles">
+                            <tree><field name="display_name"/></tree>
+                        </field>
+                    </form>`,
+                resId: 1,
+                mockRPC(route, args) {
+                    assert.step(args.method);
+                    if (args.method === "web_save") {
+                        assert.deepEqual(args.args[1], {
+                            bar: false,
+                            turtles: [[0, args.args[1].turtles[0][1], {}]],
+                        });
+                    }
+                },
+            });
+
+            assert.containsOnce(target, ".o_field_widget[name=turtles] .o_data_row");
+            assert.strictEqual(target.querySelector(".o_data_cell").innerText, "donatello");
+
+            await click(target.querySelector(".o_field_widget[name=bar] input"));
+            assert.containsN(target, ".o_field_widget[name=turtles] .o_data_row", 2);
+            assert.strictEqual(
+                target.querySelectorAll(".o_data_cell")[1].innerText,
+                "onchange name"
+            );
+
+            await clickSave(target);
+            assert.containsN(target, ".o_field_widget[name=turtles] .o_data_row", 2);
+            assert.verifySteps(["get_views", "web_read", "onchange", "web_save"]);
+        }
+    );
+
+    QUnit.test(
+        "onchange on x2many add and delete x2m record, returning to initial state",
+        async function (assert) {
+            serverData.models.turtle.fields.display_name.readonly = true;
+            serverData.models.partner.onchanges = {
+                turtles: function () {},
+            };
+
+            await makeView({
+                type: "form",
+                resModel: "partner",
+                serverData,
+                arch: `
+                    <form>
+                        <field name="turtles">
+                            <tree editable="bottom">
+                                <field name="display_name"/>
+                            </tree>
+                        </field>
+                    </form>`,
+                resId: 1,
+                mockRPC(route, args) {
+                    assert.step(args.method);
+                    if (args.method === "onchange" && args.model == 'partner') {
+                        // onchange expects a value for the modified field
+                        assert.deepEqual(Object.keys(args.args[1]), ["turtles"]);
+                        assert.deepEqual(args.args[2], ['turtles']);
```

Comment:
I would assert.deepEqual the command, to make it explicit

Reply:
Onchange is called twice, once with a create command, once with an empty command list.
So I have to somehow handle the two cases.
Tried it in a fixup commit, but dunno if it's worth it (or I'm maybe missing a magic solution way better).

Reply:
You could assert.step(JSON.stringify(args.args[1]));

Reply:
Seems unreliable as the create command contains a virtual id.
`'{"turtles":[[0,"virtual_2625",{"display_name":false}]]}'`